src/main/java/com/dj/core/model/graph/ConnectedNeuron.java

```diff
@@ -163,24 +166,18 @@ public void backwardSignalReceived(final Double error) {
         if (derivative == 0.) {
             return;
         }
-        backwardConnections.keySet().forEach(conn ->
-            backwardConnections.compute(conn, (k, weight) -> {
-                    final double newWeight = weight + inputSignals.get(conn) * dz * context.getLearningRate();
-                    if (context.isDebugMode()) {
-                        if (brokenValue(newWeight)) {
-                            throw new RuntimeException("Updated connection weight is broken");
-                        }
-                    }
-                    return newWeight;
-                }
-            ));
+
+        double dzLearningRate = dz * context.getLearningRate();
+        backwardConnections = backwardConnections.add(inputSignals.scalarMultiply(dzLearningRate));
```

Comment:
this looks awesome! especially taking into the account how many lines this change replaces:)!!!! really great job!